src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -776,33 +883,49 @@ literalChars.Length switch
+                        {
+                            writer.WriteLine("break;");
+                        }
+                        writer.WriteLine();
+                    }
+                    else
+                    {
+                        indexOfFoundBlock = EmitBlock(writer, $"if (i >= 0)");
                     }
-                    writer.WriteLine();
 
                     // We found the literal.  Walk backwards from it finding as many matches as we can against the loop.
-                    writer.WriteLine("int prev = i;");
-                    writer.WriteLine($"while ((uint)--prev < (uint)slice.Length && {MatchCharacterClass(hasTextInfo, options, "slice[prev]", target.LoopNode.Str!, caseInsensitive: false, negate: false, additionalDeclarations, requiredHelpers)});");
+                    writer.WriteLine("int prev = i - 1;");
+                    using (EmitBlock(writer, $"while ((uint)prev < (uint)slice.Length && {MatchCharacterClass(hasTextInfo, options, "slice[prev]", target.LoopNode.Str!, caseInsensitive: false, negate: false, additionalDeclarations, requiredHelpers)})"))
+                    {
+                        writer.WriteLine("prev--;");
+                    }
+                    writer.WriteLine();
 
                     if (target.LoopNode.M > 0)
                     {
                         // If we found fewer than needed, loop around to try again.  The loop doesn't overlap with the literal,
                         // so we can start from after the last place the literal matched.
-                        writer.WriteLine($"if ((i - prev - 1) < {target.LoopNode.M})");
-                        writer.WriteLine("{");
-                        writer.WriteLine("    pos += i + 1;");
-                        writer.WriteLine("    continue;");
-                        writer.WriteLine("}");
+                        using (EmitBlock(writer, $"if ((i - prev - 1) < {target.LoopNode.M})"))
+                        {
+                            writer.WriteLine("pos += i + 1;");
+                            writer.WriteLine("continue;");
+                        }
+                        writer.WriteLine();
                     }
-                    writer.WriteLine();
 
                     // We have a winner.  The starting position is just after the last position that failed to match the loop.
-                    // TODO: It'd be nice to be able to communicate i as a place the matching engine can start matching
-                    // after the loop, so that it doesn't need to re-match the loop.
+                    // We also store the position after the loop into runtrackpos (an extra, unused field on RegexRunner) in order
+                    // to communicate this position to the match algorithm such that it can skip the loop.
                     writer.WriteLine("base.runtextpos = pos + prev + 1;");
+                    writer.WriteLine("base.runtrackpos = pos + i;");
```

Comment:
Feels a bit like cheating to use this extra field here but I understand the reasons behind it. We may want to clean this up in the future and instead emit a field on the generated type to store this or to change the return of FFC to also return this optional value

Reply:
Yeah, it's a little odd, but at the and time, we have this unused field just sitting here unused, and we can't get rid of it unless we're willing to sacrifice compat, so... yeah. We also can't emit a field in RegexCompiler, though we could pass state out and in if the helpers. This was one of the things much simpler when just emmitting all the code into the one method, as then they could just refer to the same local.